### ISSUE TITLE
Plugin long option value not populated in this.options when using a shortcut option on command line

### ISSUE THREAD

**Issue by matt-miller-cargill** [serverless/serverless#9886]
Our project has created a number of custom serverless plugins that take params via the command line.  In version 1.72.0 of serverless these command line values are available from this.options in both long and shortcut forms of the option.  In version 2.56 of serverless if you use a the shorcut form of an argument, you can not access the value from this.options['myLongOption'].  With this current functionality we are not upgrading to 2.56 because we do not want to have to update all of our plugins to read values with code such as this.options['myLongOption'] || this.options['m'].

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: myservice
frameworkVersion: '2'
plugins:
    - './lib/TestPlugin'

provider:
  name: aws
  runtime: nodejs12.x
  lambdaHashingVersion: 20201221

functions:
  hello:
    handler: handler.hello
```

</details>

<details>
<summary><b><code>serverless foo -x hello</code> output</b></summary>

```
Serverless: Load command interactiveCli
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command config:tabcompletion
Serverless: Load command config:tabcompletion:install
Serverless: Load command config:tabcompletion:uninstall
Serverless: Load command create
Serverless: Load command install
Serverless: Load command package
Serverless: Load command deploy
Serverless: Load command deploy:function
Serverless: Load command deploy:list
Serverless: Load command deploy:list:functions
Serverless: Load command invoke
Serverless: Load command invoke:local
Serverless: Load command info
Serverless: Load command logs
Serverless: Load command metrics
Serverless: Load command print
Serverless: Load command remove
Serverless: Load command rollback
Serverless: Load command rollback:function
Serverless: Load command slstats
Serverless: Load command plugin
Serverless: Load command plugin
Serverless: Load command plugin:install
Serverless: Load command plugin
Serverless: Load command plugin:uninstall
Serverless: Load command plugin
Serverless: Load command plugin:list
Serverless: Load command plugin
Serverless: Load command plugin:search
Serverless: Load command config
Serverless: Load command config:credentials
Serverless: Load command upgrade
Serverless: Load command uninstall
Serverless: Load command foo
Serverless: Load command login
Serverless: Load command logout
Serverless: Load command generate-event
Serverless: Load command test
Serverless: Load command dashboard
Serverless: Load command output
Serverless: Load command output:get
Serverless: Load command output:list
Serverless: Load command param
Serverless: Load command param:get
Serverless: Load command param:list
Serverless: Load command studio
Serverless: Skipping variables resolution with old resolver (new resolver reported no more variables to resolve)
Serverless: Invoke foo
this.options={"x":"hello"}
this.options.someString=undefined
this.options.x=hello
```

</details>

<b>Installed version</b>

```
serverless --version
Framework Core: 2.56.0 (standalone)
Plugin: 5.4.4
SDK: 4.3.0
Components: 3.16.0
```

<b>TestPlugin.js</b>
```
"use strict";

class TestPlugin {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;

    this.commands = {
      foo: {
        usage: "Some Foo Usage",
        lifecycleEvents: ["functions"],
        options: {
          someString: {
            usage: "Look at a string option",
            shortcut: "x",
            required: true,
            type: "string",
          },
        },
      },
    };

    this.hooks = {
      "foo:functions": this.fooFunction.bind(this),
    };
  }

  fooFunction() {
    console.log(`this.options=${JSON.stringify(this.options)}`);
    console.log(`this.options.someString=${this.options.someString}`);
    console.log(`this.options.x=${this.options.x}`);
  }
}

module.exports = TestPlugin;
```

**Comment by matt-miller-cargill** [comment]
@medikoo we could update the plugins for sure, but we were hoping to avoid that since this seems like a breaking change.  Seems odd to make a user check both the long form and the short form because they can't be different values (at least I hope they can't).

**Comment by medikoo** [comment]
> @medikoo we could update the plugins for sure, but we were hoping to avoid that since this seems like a breaking change. 

@matt-miller-cargill why exactly do you think it would be a breaking change for you? Long form (no matter if short or long was used in terminal) works across in  published `serverless` versions

**Comment by matt-miller-cargill** [comment]
@medikoo it breaks our existing plugins that only read command line args with their long-arg-name.  To fix this issue we would have to update all of our plugins to read command line args with BOTH their long-arg-name and shortcut-name.  Currently our serverless plugins are being called with various automation scripts that make use of shortcut-names, and they do not work on the newest version of serverless.

**Comment by medikoo** [comment]
>  To fix this issue we would have to update all of our plugins to read command line args with BOTH their long-arg-name and shortcut-name

Why do you assume it's the case? Why do you think just sticking to long-arg-name version won't work?

**Comment by matt-miller-cargill** [comment]
@medikoo The attached example shows that you can't access a shortcut-name arg with the long-arg-name.

**Comment by medikoo** [comment]
@matt-miller-cargill I've doubled checked, and indeed  in case of plugin introduced options the long name is not resolved as expected from alias. That's a clear bug, we will fix it shortly

**Comment by medikoo** [comment]
@matt-miller-cargill I've looked closer, and actually issue is behind Framework design and how it deals with plugins.

Note that in your test you confirm  on `options` which are passed to Plugin constructor _before_ Framework obtains the schema from plugin about introduced commands and options.

This `options` object has no chance of being fulfilled properly, and it should not be trusted in this case.

To have a fulfilled CLI options object, best is to read fully resolved `options` in `initialize` hook, as e.g. Google cloud plugin does:
https://github.com/serverless/serverless-google-cloudfunctions/blob/390bba5bee3796969c9798629bd588917759616d/invokeLocal/googleInvokeLocal.js#L18-L20

If you add that to your plugin, you'll see option long name accessible on `this.options`

**Comment by matt-miller-cargill** [comment]
@medikoo this seems like a hacky way to do it.  
1. this.options has to be set two times which is odd (once in the constructor, then once in the initialize block.  

2. My example code came directly from the serverless docs and this is not mentioned: https://www.serverless.com/framework/docs/providers/aws/guide/plugins#defining-options 

3. It's still a breaking change that will mess up existing plugins written before this new behavior.

**Comment by medikoo** [comment]
@matt-miller-cargill I've looked into how it worked with v1.72.0 and over there same `options` object that was passed to the constructor was later internally patched to have options state updated, it's the reason why it contains expected option in hook callback.

This was lost with CLI options refactor introduced with v2.24.0, and that change can be coined as a bug. I just made a patch to ensure `options` object is patched (and not overridden) in a similar manner as it was before, So with next release that'll be fixed.

Nonetheless, any state-related data that's passed by the Framework to the plugin constructor should be considered unstable (state is not yet fulfilled at that point). It's a design issue on our side that we structured constructor to pass _options_ at this point, and that we embraced this in documentation.

Constructor should be purely about plugins providing its config to the Framework, and no state should be read at that point by the plugins/

With @mnapoli we will soon work on new better documentation for plugins authors, and we'll take that issue into account
